R/slide.R

```diff
@@ -223,6 +221,35 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
     after <- time_step(after)
   }
 
+  inrange_time_values_not_in_x = ref_time_values - before
+  inrange_time_values_not_in_x <- inrange_time_values_not_in_x[!(inrange_time_values_not_in_x %in% unique(x$time_value))]
+
+  # Do set up to let us recover `ref_time_value`s later.
+  # A helper column marking real observations.
+  x$.real = TRUE
+
+  # Create df containing phony data. Df has the same columns and attributes as
+  # `x`, but filled with `NA`s aside from grouping columns. Number of rows is
+  # equal to the number of `inrange_time_values_not_in_x` we have * the
+  # number of unique levels seen in the grouping columns.
+  before_time_values_df = data.frame(time_value=inrange_time_values_not_in_x)
+  if (length(group_vars(x)) != 0) {
```

Comment:
We have the dataframe `before_time_values_df` containing one column, `time_value`, of `n` dates.

If `x` is grouped and thus has `group_vars`, we need to duplicate the `n` dates in `before_time_values_df` for each unique combination of `group_vars`. This is what the `cross_join` is doing. This makes each of the phony obs and associated date in `before_time_values_df` available during [the `group_modify`](https://github.com/cmu-delphi/epiprocess/blob/0b48ea4a4ff18d9da2cade6470b424b1275be06e/R/slide.R#L364), since `group_modify` is splitting by group (AKA unique `group_vars` combinations).

If `x` is not grouped, we don't need to duplicate the `n` dates in `before_time_values_df`, since all phony obs and associated dates will be available during [the `group_modify`](https://github.com/cmu-delphi/epiprocess/blob/0b48ea4a4ff18d9da2cade6470b424b1275be06e/R/slide.R#L364). That is, `group_modify` isn't splitting the data into subsets.

I can get the `cross_join` to work more generally (currently doesn't work if `x` isn't grouped + a bit fiddly), but I was trying to avoid doing unnecessary work in the ungrouped case. If there are no `group_vars`, we'd end up joining `before_time_values_df` onto an empty df and expecting back the original version of `before_time_values_df`.

> what happens in the else case? I would have thought that... we need special processing

Were you thinking of something in particular?

Reply:
I might have been thinking of duplication regarding non-grouping key variables? But that's not handled in the grouped case, and not necessary in either case.

However, I am surprised how we seem to get the right recycling behavior in either case in this type of situation.  I don't see a `.real` filter being applied when determining the recyling counts.  Trying to figure out why it works.

Reply:
Ah, I see, it's also due to the first part of `slide_one_grp` which mutates the `time_values` binding.